File: balancedbrackets/prog.rb
t = gets.strip.to_i


# this works, but it's o(N*N) very slow...
#
# @pairs = { "[" => "]", "{" => "}", "(" => ")" }
# def remove_adjacents(expression)
#   expression.each_with_index do |char,i|
#     next_char = expression[i+1]
#     if next_char && next_char == @pairs[char]
#       expression.delete_at(i)
#       expression.delete_at(i)
#       expression = remove_adjacents(expression)
#     end
#   end
#   return expression
# end
#
# t.times do
#   expression = gets.strip.chars
#   balanced = true
#   l = expression.length
#   balanced = false if l%2 == 1 || remove_adjacents(expression).any?
#   puts balanced ? "YES" : "NO"
# end

# O(N) way better
def is_balanced?(expression)
  stack = []
  expression.each do |char|
    if char == '{'
      stack << '}'
    elsif char == '['
      stack << ']'
    elsif char == '('
      stack << ')'
    else
      return false if stack.empty? || char != stack.pop
    end
  end
  return stack.empty?
end

t.times { puts is_balanced?(gets.strip.chars) ? "YES" : "NO" }

File: balancedbrackets/prog.swift
import Foundation

struct Stack<Element> {
    var items = [Element]()
    mutating func push(_ item: Element) {
        items.append(item)
    }
    mutating func pop() -> Element {
        return items.removeLast()
    }
    func isEmpty() -> Bool {
      return items.isEmpty
    }
}

func is_balanced(expression: [Character]) -> Bool {
  var stack = Stack<Character>()
  for char in expression {
    if char == "{" { stack.push("}") }
    else if char == "(" { stack.push(")") }
    else if char == "[" { stack.push("]") }
    else {
      if(stack.isEmpty() || stack.pop() != char) { return false }
    }
  }
  return stack.isEmpty()
}

let n = Int(readLine()!)!
for i in 0..<n {
  let expression = Array(readLine()!.characters)
  is_balanced(expression: expression) ? print("YES") :  print("NO")
}
